sql/core/src/main/scala/org/apache/spark/sql/execution/MetadataOnlyOptimizer.scala

```diff
@@ -0,0 +1,190 @@
+
+      case relation: CatalogRelation if relation.catalogTable.partitionColumnNames.nonEmpty =>
+        (Some(relation), Seq.empty[NamedExpression], Seq.empty[Expression], Map.empty)
+
+      case p @ Project(fields, child) if fields.forall(_.deterministic) =>
+        val (plan, _, filters, aliases) = findRelation(child)
+        val substitutedFields = fields.map(substitute(aliases)).asInstanceOf[Seq[NamedExpression]]
+        (plan, substitutedFields, filters, collectAliases(substitutedFields))
+
+      case f @ Filter(condition, child) if condition.deterministic =>
+        val (plan, fields, filters, aliases) = findRelation(child)
+        val substitutedCondition = substitute(aliases)(condition)
+        (plan, fields, filters ++ Seq(substitutedCondition), aliases)
+
+      case _ => (None, Seq.empty[NamedExpression], Seq.empty[Expression], Map.empty)
+    }
+  }
+
+  private def convertToMetadataOnlyPlan(
+      parent: LogicalPlan,
+      projectList: Seq[NamedExpression],
+      filters: Seq[Expression],
+      relation: LogicalPlan): LogicalPlan = relation match {
+    case l @ LogicalRelation(files: HadoopFsRelation, _, _) =>
+      val attributeMap = l.output.map(attr => (attr.name, attr)).toMap
+      val partitionColumns = files.partitionSchema.map { field =>
+        attributeMap.getOrElse(field.name, throw new AnalysisException(
+          s"Unable to resolve ${field.name} given [${l.output.map(_.name).mkString(", ")}]"))
+      }
+      val filterColumns = filters.flatMap(_.references)
+      val projectSet = parent.references ++ AttributeSet(filterColumns)
+      if (projectSet.subsetOf(AttributeSet(partitionColumns))) {
+        val selectedPartitions = files.location.listFiles(filters)
+        val partitionValues = selectedPartitions.map(_.values)
+        val valuesRdd = sparkSession.sparkContext.parallelize(partitionValues, 1)
+        val valuesPlan = LogicalRDD(partitionColumns, valuesRdd)(sparkSession)
+        if (projectList.nonEmpty) {
+          parent.withNewChildren(Project(projectList, valuesPlan) :: Nil)
+        } else {
+          parent.withNewChildren(valuesPlan :: Nil)
+        }
+      } else {
+        parent
+      }
+
+    case relation: CatalogRelation =>
+      val attributeMap = relation.output.map(attr => (attr.name, attr)).toMap
+      val partitionColumns = relation.catalogTable.partitionColumnNames.map { column =>
+        attributeMap.getOrElse(column, throw new AnalysisException(
+          s"Unable to resolve ${column} given [${relation.output.map(_.name).mkString(", ")}]"))
+      }
+      val filterColumns = filters.flatMap(_.references)
+      val projectSet = parent.references ++ AttributeSet(filterColumns)
+      if (projectSet.subsetOf(AttributeSet(partitionColumns))) {
+        val partitionColumnDataTypes = partitionColumns.map(_.dataType)
+        val partitionValues = catalog.getPartitionsByFilter(relation.catalogTable, filters)
```

Comment:
for the sake of simplicity, how about we remove this optimization and always apply filters after getting the partition values? We can implement it in follow-ups.